Codegen updates

kodebach opened the pull request:
Contains some updates for the code-generation. For progress see #2680.

Fixes #2866

## Basics

- [x] Short descriptions should be in the release notes (added as entry in
      `doc/news/_preparation_next_release.md` which contains `_(my name)_`)
      **Please always add something to the the release notes.**
- [ ] Longer descriptions should be in documentation or in design decisions.
- [x] Describe details of how you changed the code in commit messages
      (first line should have `module: short statement` syntax)
- [x] References to issues, e.g. `close #X`, should be in the commit messages.

## Checklist

- [x] I added unit tests
- [x] I ran all tests locally and everything went fine
- [x] affected documentation is fixed
- [x] I added code comments, logging, and assertions (see [Coding Guidelines](https://master.libelektra.org/doc/CODING.md))
- [x] meta data is updated (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))

## Review

Reviewers will usually check the following:

- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)
- [ ] Documentation is concise and good to read
- [ ] Examples are well chosen and understandable

kodebach commented:
@markus2330 the documentation is now finished, there are still a few minor issue I want to fix, but in the meantime please review the documentation

Once everything is ready to merge, I will rename the template from `elektra` to `highlevel` as requested (the documentation should already reflect this). Renaming will be the last step, because that will probably mess with comments on github.

markus2330 commented on doc/dev/code-generator.md:
```diff
@@ -0,0 +1,211 @@
+
+Since we need to some way of supplying data to our template, we have to create a subclass of `GenTemplate`.
+
+First we create a new directory for our template class in `src/tools/kdb/gen`, for our template it should be `src/tools/kdb/gen/example`. In
```
The directory can not be somewhere else?

kodebach replied:
No, because the CMake scripts expects the files in this place.

markus2330 replied:
Are the templates then also built-in the binaries?

kodebach replied:
Which binaries?

Right now CMake reads all the `.mustache` and puts their contents into string constants (`static const char *`), because that was the easiest way to use the mustache library. We could also install the template files and read them at runtime, but then we would have to do some trickery to allow running the code-generator without installing first.

markus2330 replied:
> Which binaries?

`kdb` (and `kdb-full`, ...)

> Right now CMake reads all the .mustache and puts their contents into string constants 

Would be good to mention this here. Then it is less surprising why we modify files in `src/tools/kdb/gen`

kodebach replied:
I just noticed that I confuse the lines here. For the `.cpp` and `.hpp` files, it doesn't really matter where they are, but putting the files in a different folder might required modifying the CMake files.

I will add some clarifications.

markus2330 commented on doc/dev/code-generator.md:
```diff
@@ -0,0 +1,211 @@
+
+First we create a new directory for our template class in `src/tools/kdb/gen`, for our template it should be `src/tools/kdb/gen/example`. In
+this directory we then create `example.cpp` and `example.hpp`. If your template becomes sufficiently complex, it may make sense to split it
+into multiple files, for this reason we recommend creating a new directory for each template.
```
Is this splitup what you before named "parts"? If yes, please stick to the terms. Otherwise explain the difference.

kodebach replied:
No, here I am talking about splitting the C++ code into multiple files. I will clarify that.

markus2330 commented on doc/dev/code-generator.md:
```diff
@@ -0,0 +1,211 @@
+#ifndef ELEKTRA_EXAMPLE_HPP // choose a unique header guard
+#define ELEKTRA_EXAMPLE_HPP
+
+#include "../template.hpp"
```
Relative includes should be avoided.

kodebach replied:
Is there any particular reason for that?

markus2330 replied:
It is more clear if the different include files are in the same directory. Isn't this possible in this situation?

kodebach replied:
It should be possible, I will update the guide and move the header of the existing template, when I rename it.

kodebach replied:
I just saw that `src/tools/kdb` part of the include path, so I changed it to `#include <gen/template.hpp>`, which is a better solution anyway.

markus2330 commented on doc/dev/code-generator.md:
```diff
@@ -0,0 +1,211 @@
+all templates. Let's dissect this line:
+
+- `ExampleGenTemplate ()` we declare a zero argument constructor. All templates _must_ have only a single zero argument constructor,
+  otherwise the framework cannot instantiate them.
```
Which framework?

kodebach replied:
The code-generator framework. I called the internal setup that invokes the mustache library "framework" to distinguish it from the command line interface implemented in `gen.cpp`.

markus2330 replied:
Then describe this when introducing the terms.

kodebach replied:
It is actually explained in this paragraph at the start of the Basics section:

> The file [`src/tools/kdb/gen.cpp`](/src/tools/kdb/gen.cpp) implements the command-line interface (CLI) and is of little interest to this guide. Instead we will focus on the framework that is invoked via the CLI. The bulk of the framework is implemented in the classes [`GenTemplate` and `GenTemplateList`](/src/tools/kdb/gen/template.hpp) (and of course kainjow's mustache library).

I will add further clarification.

markus2330 commented on doc/dev/code-generator.md:
```diff
@@ -0,0 +1,211 @@
+The framework will invoke `getTemplateData` for each part of our template with the `outputName` and `parentKey` as given on the command line,
+as well as the current part suffix (`part`) and the input keyset `ks`. All keys of the input keyset are guaranteed to be below `parentKey`.
+
+How exactly the above code works is left as an exercise to the reader.
```
Explain or keep simpler.

kodebach replied:
It won't get much simpler than iterating through a `KeySet` simply collecting the key names.

markus2330 commented on doc/dev/code-generator.md:
```diff
@@ -0,0 +1,211 @@
+## Creating a new template
+
+In this guide we will create a basic template, that generates a single file containing. The file will simple list all keys in our input
+keyset. An example output would be (the same as `kdb ls`):
```
output of the template engine or of an execution of the binary from the compiled code we got from the template engine?

kodebach replied:
There is no compiled binary. We do not create a template for C code, but just a very basic text file template to explain how the code-generator framework works.

markus2330 commented on doc/dev/code-generator.md:
````diff
@@ -0,0 +1,211 @@
+Now you should be able to use the new template by running (after compiling/installing Elektra again):
+
+```
+kdb gen example user userkeys
````
In which folder are we now?

kodebach replied:
Doesn't matter. The command should be equivalent to `kdb ls user > userkeys.txt`, which also doesn't care about cwd.

markus2330 commented on doc/dev/code-generator.md:
```diff
@@ -0,0 +1,211 @@
+To access the parameter value call one of the `getParameter` overloads. One takes a default value, the other takes a map of values and
+verifies that one of the given values has been chosen. For more information see the relevant code documentation. There is also
+`getBoolParameter` which a specialised version for boolean parameters. It accepts only `0` and `1` as values.
+
```
Can you give an example how to execute `kdb gen` here?

markus2330 commented on doc/dev/code-generator.md:
```diff
@@ -0,0 +1,211 @@
+
+### Custom escape functions
+
+By default mustache escapes values for use in HTML (unless `{{{ name }}}` or `{{& name }}` is used). Since most of our templates are not in
```
are not HTML

markus2330 commented on doc/dev/code-generator.md:
```diff
@@ -0,0 +1,211 @@
+
+By default mustache escapes values for use in HTML (unless `{{{ name }}}` or `{{& name }}` is used). Since most of our templates are not in
+fact HTML, the escape function can be customised. You simply have to override `GenTemplate::escapeFunction`. For an example see
+`ElektraGenTemplate::escapeFunction`, it is designed for C code instead of HTML.
```
In which file?

kodebach replied:
`elektragen.cpp`, I expected that its fully qualified name would be enough to find the function.

markus2330 replied:
Of course you know where it is but the reader might need to use `find`.

markus2330 commented on doc/help/kdb-gen-highlevel.md:
```diff
@@ -0,0 +1,101 @@
+- `genSetters`:
+  Switches whether setters should be generated at all; allowed values: `true` (default), `false`
+- `specLocation`:
+  Changes where the processed specification will be stored; allowed values: `embedded` (default), `external`.
```
This is not clear. What is a "processed specification"?

kodebach replied:
I added an explanation to `elektra-highlevel-gen.md`. The description here is intentionally incomplete, so that people read the full documentation and don't make assumptions based on the short descriptions.

markus2330 commented on doc/help/elektra-highlevel-gen.md:
```diff
@@ -0,0 +1,413 @@
+If you don't care that additional files will be required to run your application, you can switch to `specLocation=external`. With this option
+the code-generator creates an additional output file, which contains the processed specification. The files is in `quickdump` format, so it
+can be mounted directly via `quickdump`. If embedding the specification into your application is not something you want to do (e.g. because
+of the size increase), but you still want to use `specload`, you can do that too. Just use a configuration like
```
configuration to what?

kodebach replied:
specload

markus2330 commented on doc/help/elektra-highlevel-gen.md:
```diff
@@ -0,0 +1,413 @@
+`app=/usr/bin/cat args=#0 args/#0="path-to-spec-output-file"`.
+
+Using the combination `specLocation=external defaultsHandling=embedded`, still embeds a smaller keyset only containing the default values
+into your application. If that is still too much for you, may use `defaultsHandling=embedded`. This way defaults are not passed to
```
defaultsHandling=embedded already used before?

kodebach replied:
Copy-paste error, it should obviously be `defaultsHandling=speconly` (the other setting for `defaultsHandling`)

markus2330 replied:
Was not clear to me. You need to be very careful in these explanations as this configuration is unfortunately very complicated.

kodebach commented:
> in elektra-highlevel-gen.md I read only the beginning.

Did you read kdb-gen.md and kdb-gen-highlevel.md first?

kodebach commented on doc/dev/code-generator.md:
```diff
@@ -0,0 +1,211 @@
+# Code-generator
+
+This guide focuses on writing new templates for the code-generator. For using the code generator take a look at the man-page
```
Yes, now that tutorials has a subsection for Elektra developers.

kodebach replied:
I will move the file once you are satisfied with everything else, otherwise it will probably mess up the comments on github.

kodebach commented on doc/dev/code-generator.md:
```diff
@@ -0,0 +1,211 @@
+First we need to define a few terms:
+
+- _template name_: This is the main identifier of a template. It is used in the command line to choose the template.
+- _template base name_: Unlike the template name, this identifier is purely internal. It is the base name of all mustache files belonging
```
This internal as in "the user of `kdb gen` won't ever see this, but the person writing the template still needs to now about this term.

markus2330 replied:
Better to avoid "internal" then. Use something like "only relevant for the person writing the template"

markus2330 commented:
>  Did you read kdb-gen.md and kdb-gen-highlevel.md first?

Please write in the beginning of the documents what you expect the reader to have read before and at the end of tutorials what to read next.

kodebach commented:
> Please write in the beginning of the documents what you expect the reader to have read before

https://github.com/kodebach/libelektra/blob/9617a4d47afb4f670f34f8b9f9ae980fce74ed27/doc/dev/code-generator.md#L3-L4

markus2330 commented:
The link does not work. If it is already done, good. Then I will hopefully read everything in the right way in the next review.

kodebach commented:
> The link does not work. If it is already done, good. Then I will hopefully read everything in the right way in the next review.

It seems links to markdown files can't reference line numbers...

In any case the relevant part of the file is right at the beginning:

```markdown
For the basic features see [`kdb-gen-highlevel(1)`](kdb-gen-highlevel.md)
```

`kdb-gen-highlevel.md` also refers to `elektra-highlevel-gen.md` at appropriate locations.

markus2330 commented:
> For the basic features see

If you say it like this, I would not assume that I **need to** read it first. I would still consider that I can read this document by itself.

To make sure that readers read the other document first, write something "We assume you read ..., where the more basic features are explained".

Maybe a line break before the sentence is already enough to make it more visible.

kodebach commented:
@markus2330 Do you have any other comments on this PR? If not, I would do the renames and this PR should be "ready to merge", apart from the error caused by symbol versioning.

markus2330 commented:
I'll do one more review (and also try out what is written in the tutorials) but I prefer if you do the renames now (before the review). It does not matter if some conversations get lost as you are very careful in fixing review comments. But in renames it is always possible that some place is forgotten.

markus2330 commented on doc/help/kdb-gen-highlevel.md:
```diff
@@ -0,0 +1,101 @@
+
+However, it is not recommended to have the code-generator read from the KDB, so one should instead use:
+
+`kdb gen -F ni=spec.ini highlevel /sw/org/app/#0/current config`
```
I tried to execute that (or to be more specific: `kdb gen -F ni=./examples/highlevel/spec.ini highleve /sw/org/app/#0/current config`) I get the error:

```
The command kdb gen terminated unsuccessfully with the info:
parentKey has to be in spec namespace
```

Actually, I think the command is correct, and the spec can simply be appended internally.


And quite weird: I get no (error) output when I mistype "highlevel".

When I strace the command, I get:
```
openat(AT_FDCWD, "./examples/highlevel/spec.ini", O_RDONLY) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=233, ...}) = 0
read(3, "[]\nmountpoint = highlevelexample"..., 4096) = 233
read(3, "", 4096)                       = 0
close(3)                                = 0
munmap(0x7f83a5a2b000, 2130336)         = 0
exit_group(0)                           = ?
+++ exited with 0 +++
```

So it seems like to successfully open the file, but does not do anything. It would be good to get errors if I did something wrong.

When I change the command with a key starting in spec namespace, I get a config.c and config.h

In the generated code, most of the Elektra in the comments are written incorrectly "elektra".

I'll continue with the review tomorrow.

markus2330 commented on doc/help/kdb-gen.md:
```diff
@@ -0,0 +1,71 @@
+
+## DESCRIPTION
+
+This command invokes Elektra code-generator.
```
```suggestion
This command invokes Elektra's code-generator.
```

markus2330 commented on doc/help/kdb-gen.md:
```diff
@@ -0,0 +1,71 @@
+
+`kdb gen [options...] <templateName> <parentKey> <outputName> [parameters...]`
+
+- `[options...]`:
```
This is now inconsistent with the other man pages (where the options are not mentioned here).

markus2330 commented on doc/help/elektra-highlevel-gen.md:
```diff
@@ -0,0 +1,415 @@
+
+The parameters that are relevant to the concepts described here are (for the rest see [`kdb-gen-highlevel(1)](kdb-gen-highlevel.md)):
+
+- `enumConv`: allowed values: `strcmp`, `trie`, `default` (default)
```
And the enumConf option is not needed if the default works properly. Did you benchmark if there are really scenarios where the trie is noticeable slower?

kodebach replied:
> the enumConf option is not needed if the default works properly

The option is already there and it has basically no impact on the code (of the code-generator template). It also doesn't complicate the documentation, since even without the option, we would need to explain the two different outcomes of using enums. So I don't really see any reason to remove the option.

> Did you benchmark if there are really scenarios where the trie is noticeable slower?

I only did very quick tests. As far as I can tell, the trie implementation is quicker even if the enum values match the regexes `a{31}a` and `a{31}b`. But the runtime of the trie version seems to increase quicker than the strcmp version (with respect to the string length).

However, we are talking about a difference of a few hundred milliseconds per 10.000.000 executions, so it hardly matters. (Also my CPU is quite old, newer CPUs might perform better in strcmp.)

The main reason why I didn't use the trie version all the time is, that is becomes unreadable very very quickly. That is also why the option exists. If you want optimal performance `trie` *might* be better, for readability `strcmp` is probably better.

markus2330 replied:
> we would need to explain the two different outcomes of using enums.

If we do not have the option we would not need to explain anything. But you already gave a good reason why we need the option.

> The main reason why I didn't use the trie version all the time is, that is becomes unreadable very very quickly. That is also why the option exists.

Then say exactly this. Why bringing up wild speculations of performance if this option is actually about being able to understand the output of the generator. (Which is a perfectly good reason to have the option.)

markus2330 commented on doc/help/elektra-highlevel-gen.md:
```diff
@@ -0,0 +1,415 @@
+
+The `enumConv` option is described [below](#enums).
+
+With `specLocation` and `defaultsHandling` you can configure what data is embedded into your application's binary. The default option for
```
I am still not happy with this here. You describe a few scenarios but only write little about what these options actually do. Furthermore you describe the two options together as if they were one option. The tutorial also makes very clear (what was to me clear before) that it would be much better if we simply have one option for these selected scenarios.

But I also see that you really tried hard to make the tutorial nice, so if it is too much for you, you can leave it. I'll then read the docu again and try to give you better input how to describe the config options. The tutorial suggests to use these options like a single option anyway.

kodebach replied:
> but only write little about what these options actually do.

I described everything these options do. `specLocation` switches between the embedded spec and the additional quickdump file. `defaultsHandling` enables/disables embedded default values.

> The tutorial also makes very clear (what was to me clear before) that it would be much better if we simply have one option for these selected scenarios.

I will see, if I find a sensible way to unify them.

markus2330 replied:
> I described everything these options do. specLocation switches between the embedded spec and the additional quickdump file. defaultsHandling enables/disables embedded default values.

These sentence are quite descriptive but you need to clarify "embedded spec" and "default values". Ideally this should be the same ("default values" a subset of "embedded spec").

> I will see, if I find a sensible way to unify them.

:sparkling_heart:

kodebach commented on doc/tutorials/highlevel.md:
````diff
@@ -0,0 +1,322 @@
+/* elektra* macros ... */
+```
+
+At the top of the file (after some helper macros) we can see all the `ELEKTRA_TAG_*` macros. We use these to allow for some flexibility in
````
> The essence

I wouldn't call the `ELEKTRA_TAG_*` macros the essence. You could easily ignore them. The actual important part are all the `static inline` functions.

markus2330 replied:
Please feel free to use your own words, it was only a suggestion.

"the essence" in the sense that this is the trick how the API can be type safe. Or did you change that?

kodebach replied:
> Or did you change that?

Yes, the tag macros are just aliases. What makes the whole thing type safe is that each key gets its own `static inline` function.

markus2330 replied:
These problems (what is more "important") should go away anyway when you split up the snippets into several parts and explain each part individually.

kodebach commented on doc/tutorials/highlevel.md:
````diff
@@ -0,0 +1,322 @@
+
+```c
+int main (int argc, const char ** argv) {
+    specloadCheck (argc, argv);
````
It allows a bit more flexibility. Some applications might not want to call `loadConfiguration` immediately after entering `main`.

markus2330 replied:
Yes, exactly. This flexibility is what I do not want here: There are many ways how you could call it wrongly or not call at all. This is basically also a configuration option (like the others in the generator we discussed so much),  only that it is hard-coded. The less such options we have, the more streamlined the behavior of applications using Elektra is.

markus2330 replied:
> Some applications might not want to call loadConfiguration immediately after entering main.

If they don't want to, they most likely also cannot call specloadCheck directly after main. A concept which requires to do something in main is fundamentally broken.

kodebach replied:
> If they don't want to, they most likely also cannot call specloadCheck directly after main.

`specloadCheck` is very lightweight. It starts with

```
if (argc != 2 || strcmp (argv[1], "--elektra-spec") != 0)
{
    return;
}
```

so it either returns more or less immediately, or it doesn't return at all (as expected in specload mode). If it were integrated in `loadConfiguration` on the other hand, we either don't return (specload mode) or we execute the whole Elektra initialization immediately.

Applications may want to start some other work (independent of the configuration) ASAP (e.g. connect to a server, open a log file, create some worker threads). Doing that before `specloadCheck` isn't wise, because it may not return and resources may leak. On the other hand doing it after `loadConfiguration` may cause unwanted start-up delays.

> A concept which requires to do something in main is fundamentally broken.

Yes, we already discovered that the concept of `specload` is not the best.

markus2330 replied:
Yes, I am afraid you are right. There will be situations where `loadConfiguration` is called too late.  E.g. applications might modify the output streams and garbage gets printed (or nothing) instead of the specification. But even if someone needs to calls it earlier, we still can call it within `loadConfiguration`, as it can never harm to be called again later.

So we should:

- call it within `loadConfiguration` which should cover 99% of the cases
- `specloadCheck` should be renamed: it does not only check but also do it, e.g. `exitForSpec`?
- for the remaining 1%, we offer a `exitForSpec` but at the end of the tutorial with clear statements when it might be needed (and that it is usually not needed. The cases you gave usually need configuration: to which server to connect, which log file to open and how many working threads to open. But of course it is possible that some applications do this without any configuration from Elektra)

kodebach replied:
I renamed it to `exitForSpecload`, but I did not integrate it into `loadConfiguration`.

The reason is simple. To call `exitForSpecload` you need to pass `argc` and `argv`. That means you need to pass those to `loadConfiguration` as well. This is no problem, if you don't use `exitForSpecload` and call `loadConfiguration` at the start of your `main` function.

But if for some reason you used `exitForSpecload` the best way we could allow calls to `loadConfiguration` is `loadConfiguration(elektra, error, 0, NULL);`. The two extra arguments are just clutter and don't achieve anything. There are also cases (e.g. threaded applications) where `loadConfiguration` needs to be called multiple times, but `exitForSpecload` must only be called once.

Also the documentation would not become simpler. We would still have to explain, why you have to call `loadConfiguration` at the start of your `main` before printing anything to `stdout` and that any resources acquired before calling `loadConfiguration` can only be cleaned up via `atexit`, because `loadConfiguration` may call `exit`.

All of that leads me to the conclusion that it is much better to keep the functions separate.

markus2330 replied:
Yes, of course: it would not be a good solution to add argc, argv to loadConfiguration. I was thinking about some solution where this is not needed, like with support from the global plugin gopts (loadConfiguration could add the spec to "spec" and gopts sends whatever is in "spec/"?)?

The problem is that exitForSpecload is a specific hack which is only useful for the specload approach. So when an application developer thinks, he/she does not need the specload feature, he/she will not add the exitForSpecload. And this is a pity in the case that the specification is actually built-in. If we somehow can handle the specload case without forcing the user to add some additional code, the API would be more simple and all applications would be specload-aware.

kodebach replied:
`exitForSpecload` is only generated, if the specification is built-in. So if the function exists, it must be called for correct behaviour. If it doesn't exist, its obvious that we don't call it.

> like with support from the global plugin gopts

Technically this would be possible, although a bit complicated. However, I don't like the solution. First of all it is again far to much magic for my taste. Especially in a language like C, I want to know what exactly all of the code does. Otherwise it is far to easy to create serious flaws.

Right now the fact that `exitForSpecload` expects `argc` and `argv` makes it pretty obvious that it has to be called in `main`. It might still be called to late, but that cannot be prevented (except with good documentation).

Without the explicit arguments, it would be even easier to miss the very strict requirements for the function handling `specload` (whichever function that may be).

kodebach replied:
One option would be to have a static field that is set by the `exitForSpecload` and checked by `loadConfiguration`. If it the field is not set when `loadConfiguration` is called, we immediately return an error.

markus2330 replied:
> exitForSpecload is only generated, if the specification is built-in. So if the function exists, it must be called for correct behaviour. If it doesn't exist, its obvious that we don't call it.

This makes even more a cliff between specload-aware applications and others that are not. I think you should always generate the function, even if it does nothing. Then making an application specload-aware is only a build-option and not a source-code change (if exitForSpecload was added correctly already).

> One option would be to have a static field that is set by

I do not like the idea of enforcing  something that is actually a problem in the API, it only creates even more problems (e.g. plugins using loadConfiguration would fail if used from an application that does not call exitForSpecload).

Furthermore, to do it properly we would need pthreads for a mutex (otherwise we would produce potential race conditions).

> First of all it is again far to much magic for my taste.

I understand the concerns that it is kind of "magical" what gopts does in that case of exitForSpecload but gopts is already past the point of being easily understandable. But in any case it is understandable for people that are interested, adding one more if for exitForSpecload would not make a big difference.

But of course at some point the line needs to be drawn, e.g., the next step would be shouldn't we also add printHelpMessage within gopts by default to get uniform behavior between applications?

Furthermore, a big point against exitForSpecload integrated in gopts is that plugins in Elektra actually should not call exit.

Maybe you can write down what we discussed here in an decision? (So that there is some rationale for exitForSpecload, and maybe you find one more point for/against it in the process.)

kodebach replied:
> Then making an application specload-aware is only a build-option and not a source-code change (if exitForSpecload was added correctly already).

If someone wants an application that supports both embedded and external spec, they can always add an `#ifdef` with a macro injected by the build system.

I think it is much more confusing to have an `exitForSpecload` function, when `specload` is not used.

> already past the point of being easily understandable

`gopts` is very easy to understand. It uses system specific APIs to retrieve `argc`, `argv` and `envp` and then calls `elektraGetOpts`. To understand the system specific APIs you obviously need some understanding of the system. But even then apart from win32 the implementations are pretty simple.

>  the next step would be shouldn't we also add printHelpMessage within gopts by default to get uniform behavior between applications?

What do you mean by that?

> Furthermore, a big point against exitForSpecload integrated in gopts is that plugins in Elektra actually should not call exit.

When did we start talking about putting `exitForSpecload` into `gopts`? That makes no sense whatsoever. `gopts` is completely independent of any code-generated functions. The name `exitForSpecload` can also be changed by the user to match possible naming conventions.

markus2330 replied:
>  If someone wants an application that supports both embedded and external spec, they can always add an #ifdef with a macro injected by the build system.

I do not like the idea that one needs to support some kind of spec by changing code. The code should always be the same, only the options to the code generator should choose what is supported.

> I think it is much more confusing to have an exitForSpecload function, when specload is not used.

As application developer you do not know if specload will be used by the distributions. The whole point of Elektra is to move such decisions to distributions/end users. Think of specload as yet-another configuration-file format.

> When did we start talking about putting exitForSpecload into gopts?

Integrating does not necessarily mean to put it there. Of course it would be decoupled so that gopts would simply output whatsoever is in spec.

> The name exitForSpecload can also be changed by the user to match possible naming conventions.

I am also not so thrilled about that feature as it does not really offer what someone would expect. The non-generated high-level API would still not match. But this feature does not look harmful.

kodebach replied:
> The code should always be the same, only the options to the code generator should choose what is supported.

This does work here, but is not a good approach in general. It is a **code** generator. One has to expect that changing its arguments might require changes in other code.

> As application developer you do not know if specload will be used by the distributions.

The decision here is not whether to use `specload` or not, but whether to embed the specification. `specload` can be used without embedding the specification and this is also mentioned in the documentation.

> I am also not so thrilled about that feature as it does not really offer what someone would expect. The non-generated high-level API would still not match. But this feature does not look harmful.

The difference here is the non-generated API actually is an API and delivered as a shared library. It is completely external from the point of an application developer. The code-generated files are delivered as source-code and are part of the application itself.

Also, if we were to remove the renaming feature, we would need to add some `elektra` prefix to the names to avoid name conflicts that way.

---

To some this discussion up:

- I will change the templates, so that `exitForSpecload` is always generated. Although I still find it weird to generate empty and useless functions.
- I will try to sum up this discussion in a decision.

markus2330 replied:
Thank you! :sparkle:

kodebach commented on doc/tutorials/code-generator.md:
````diff
@@ -0,0 +1,240 @@
+For our example we will use this implementation:
+
+```cpp
+kainjow::mustache::data ExampleGenTemplate::getTemplateData (const std::string & outputName, const std::string & part,
````
The warnings are to be expected. I don't think it is the aim of this guide to explain how to avoid these warnings. I also think it is not wise to add `ELEKTRA_UNUSED` to the code in the example. If we did, people probably would forget to remove them, if the do use the parameter.

markus2330 replied:
Yes, the warnings is a minor problem, the header file is important.

kodebach commented on doc/help/elektra-highlevel-gen.md:
```diff
@@ -0,0 +1,415 @@
+
+This is not only hard to read, but also not very efficient. Of course we could just look at `string[1]` and `string[4]` or at
+`strlen(string)` or at any number of other things. But writing a code-generator that does this for the general case is very hard. That is
+why the default option `enumConv=default` uses the trie version, if the depth is less than 3 (i.e. if looking at `string[0]` and `string[1]`
```
No, the trie implementation is very basic. Depth `n` means looking at the first `n` characters `string[0], string[1], ..., string[n-1]`.

markus2330 replied:
Yes, then clarify in text. Maybe the compiler optimizes this anyway, so it is not so clear how efficient it is.

Btw.: I would not call this trie. Trie is a data structure and not this arrangement of switch/case statements. So better call it switch vs. strcmp. (Or leave the whole option away if you find out what is faster.)

kodebach replied:
> Btw.: I would not call this trie. Trie is a data structure and not this arrangement of switch/case statements.

I called it trie, because internally an actual trie is used to generate the switch/case statements.

> So better call it switch vs. strcmp.

Will do.

kodebach commented on doc/help/elektra-highlevel-gen.md:
```diff
@@ -0,0 +1,415 @@
+cause the initialization function (by default named `loadConfiguration`) to fail, if the specification is not mounted at the expected
+mountpoint or if the specification was not `spec-mount`ed.
+
+<sup id="f1">1</sup>: At the time of writing, the processed specification is the same as the input specification, but in future the code-generator may
```
HTML code for superscript. It produces a <sup>1</sup>. It is a bit of a hack to get a footnote.

kodebach commented on doc/help/elektra-highlevel-gen.md:
```diff
@@ -0,0 +1,415 @@
+mountpoint or if the specification was not `spec-mount`ed.
+
+<sup id="f1">1</sup>: At the time of writing, the processed specification is the same as the input specification, but in future the code-generator may
+automatically add metadata, e.g. to allow better validation. [↩](#a1)
```
I'm not sure what you want to say here...

markus2330 replied:
That you should not make this distinction, it only confuses. It is already complicated enough that there is:

1. an internal spec
2. an internal spec that might consist only of default values
3. an external (mounted) spec

Here some further explanations would be important. The "processed" specification should only be explained in the moment when it actually exists.

kodebach commented on doc/tutorials/highlevel.md:
````diff
@@ -0,0 +1,322 @@
+
+To explain how to use the generated code, lets take a look at some of `conf.h`.
+
+```c
````
Do you mean the code snippet? It is already shortened, but I will try to shorten it a bit more.

markus2330 replied:
Yes, I mean the snippet together with the text below but I was more thinking of a split-up of the snippet into small pieces with a short explanation text below each snippet.

Having the whole code in one place within the tutorial does not have any benefit, as the user already generated the whole file conf.h anyway (and can a look there to see the whole picture).

kodebach commented on doc/tutorials/highlevel.md:
````diff
@@ -0,0 +1,322 @@
+kdb_double_t mydouble = ELEKTRA_GET (Mydouble) (elektra);
+```
+
+Alternatively, you can use the convenience macro `elektraGet` and/or the tag macros. We recommend you use one of the ways below, to avoid
````
I will remove `ELEKTRA_GET (Mydouble) (...)`, if you want to (although it explains what the tag macros actually do).

I will not remove `ELEKTRA_GET (ELEKTRA_TAG_MYDOUBLE) (...)`. Yes it is more verbose than `elektraGet (...)`, you cannot get a function pointer via the `elektraGet` macro.

markus2330 replied:
It is okay to (at one point) explain how it works. But here we already switched to the user-view. And the user of the API should not use `ELEKTRA_GET (Mydouble) (elektra)`.

markus2330 replied:
Btw. Why would you need a function pointer? Aren't such situations better covered with the internalnotification?

kodebach replied:
> the user of the API should not use `ELEKTRA_GET (Mydouble) (elektra)`

Like I said, I will remove  `ELEKTRA_GET (Mydouble) (elektra)`.
But I see no technical reason to discourage  `ELEKTRA_GET (ELEKTRA_TAG_MYDOUBLE) (elektra)`. Difference between `ELEKTRA_GET (ELEKTRA_TAG_MYDOUBLE) (elektra)` and `elektraGet (elektra, ELEKTRA_TAG_MYDOUBLE)` is purely personal preference.

> Why would you need a function pointer?

I don't know, but I'm sure somebody somewhere has a use case. Mostly I don't see a reason to hide `ELEKTRA_GET (...) (...)` in the documentation. It is immediately obvious from looking at either the definition of `elektraGet` or the signature of any of the `static inline` functions that this will work. I also dislike APIs/libraries that pretend to be magic, when in fact they are not complicated at all and could easily explain how everything works.

> Aren't such situations better covered with the internalnotification?

Depends what you do with the function pointer. Also the internalnotification stuff is not supported via the highlevel API right now.

markus2330 replied:
> But I see no technical reason to discourage [...] Mostly I don't see a reason to hide ELEKTRA_GET (...) (...) in the documentation

It is only about if we really want to maintain two different kind of APIs in one API.

> I also dislike APIs/libraries that pretend to be magic, when in fact they are not complicated at all and could easily explain how everything works.

You are completely right here. It is very good that you explain. But it is something different if we also give guarantees about API stability.

> Depends what you do with the function pointer. Also the internalnotification stuff is not supported via the highlevel API right now.

Hopefully they can coexist peacefully in the same application. It is not tested, though.

kodebach commented on doc/tutorials/highlevel.md:
````diff
@@ -0,0 +1,322 @@
+problems in case we have to change how we generate the names of the `static inline` functions.
+
+```c
+kdb_double_t mydouble = ELEKTRA_GET (ELEKTRA_TAG_MYDOUBLE) (elektra);
````
see above

kodebach commented on doc/tutorials/highlevel.md:
````diff
@@ -0,0 +1,322 @@
+    elektraClose (elektra);
+    exit (EXIT_SUCCESS);
+}
+```
````
> Please move the description of how to compile here.

Why would you want to compile at this point, the application doesn't do anything useful yet. If you follow the guide everything is explained eventually.

markus2330 replied:
Seems like I was not patient enough. I think it is good to have a working example asap (aka "hello world"). So compilation should be already at this point.

kodebach replied:
IMO the "Hello, World!" of Elektra is reading a config value, but I will add a note about compiling here.

markus2330 replied:
Yes, I agree. Maybe you already put one elektraGetString in this initial example?

kodebach commented on doc/tutorials/highlevel.md:
````diff
@@ -0,0 +1,322 @@
+}
+```
+
+Admittedly writing is not quite as easy as reading, because of the error handling, but that is just a necessary evil when working in C. It
````
Well C **is** the reason that you need to have at least 5 lines of code to make a setter call safe. There are no exceptions and using return values instead of pointer arguments wouldn't change much. But I guess we don't need to mention it here, so I will remove it.

markus2330 replied:
Yes, my comment is about if the sentence contributes to the goal of the tutorial, not if it is wrong.

kodebach commented on doc/tutorials/highlevel.md:
```diff
@@ -0,0 +1,322 @@
+
+## Running your application
+
+Running your application is easy, just run the executable (e.g. `myapp`). While this might work out of the box, you will just get the default
```
For the application written in this guide it should work. It will not work, if there are no embedded defaults or if validation of the mountpoint is enabled and the mountpoint isn't there.

I will look into your error.

markus2330 replied:
I tried to stick to what is in the tutorial. In any case, you should clarify when it works, the "might" is very weak there.

kodebach commented on doc/tutorials/highlevel.md:
```diff
@@ -0,0 +1,322 @@
+float myfloat1 = ELEKTRA_GET (ELEKTRA_TAG_MYFLOATARRAY) (elektra, 1);
+// or
+kdb_float_t myfloat0 = elektraGetV (elektra, ELEKTRA_TAG_MYFLOATARRAY, 0);
+kdb_float_t myfloat1 = elektraGetV (elektra, ELEKTRA_TAG_MYFLOATARRAY, 1);
```
Is everything mounted and spec-mounted correctly? It seems to me that the `spec` plugin wasn't executed, or it didn't receive the specification.

markus2330 replied:
Again, I tried to stick to the tutorial. Afair everything was fine, and the defaults on the non-arrays also worked perfectly. If there is a difference between arrays and non-arrays, it should be noted in the tutorial.

kodebach replied:
Defaults for arrays only work via the `spec` plugin. I added a note under "Writing a specification".

kodebach commented on doc/tutorials/code-generator.md:
```diff
@@ -0,0 +1,240 @@
+[here](https://mustache.github.io/mustache.5.html). All of features supported by the kainjow library should be supported by our framework
+as well.
+
+Our CMke script will collect all `.mustache` files in `src/tools/kdb/gen/templates` into a header containing a `static const char *` field
```
> more than the template is needed.

What do you mean with "more than the template"?

markus2330 replied:
It is about giving the reader of the tutorial always a good overview of what will happen. If you write about that the CMake script collects the .mustache files, the reader might wrongly think everything is within the template.

So some introduction sentence of about what will happen (we will write a template, then we will integrate the template with C++ code, ...) is missing.

kodebach commented on doc/tutorials/highlevel.md:
````diff
@@ -0,0 +1,322 @@
+the KDB, similar to how mounting an external hard drive informs the OS about a new part of the file system.
+
+```sh
+sudo kdb mount -R noresolver myapp_spec.eqd "spec/sw/example/myapp/#0/current" specload app="$PWD/myapp"
````
`myapp_spec.eqd` is the override file that contains the changes made by the user and validated by `specload`. The line above would create a file in the current working directory whenever you call `kdb set spec/sw/example/myapp/#0/current/...`.

This line is obviously far from ideal. In a perfect world it would be `sudo kdb mount myapp_spec.eqd "spec/sw/example/myapp/#0/current" specload app="$PWD/myapp"`. But there are some problems in `specload`, if a real resolver is used. Mainly that `specload` won't be called, because the resolver reports no changes.

If you want to make changes to the spec via `specload`, you should provide an absolute path on the system level (e.g. `/etc/myap_spec.eqd`), so that we always access the same file.

markus2330 replied:
Yes, maybe we should give it a full path and write down that this file will be used for user-changes.

kodebach commented:
@markus2330 I resolved all the comments while implementing them. If anything isn't done correctly, feel free to unresolve the conversation.

kodebach commented:
@markus2330 I added a few new things to this PR:

- The file `kdbconfig.h` is no longer installed. It was causing problems with the `config.h` file from LCDproc. This required a few changes to plugins, bindings, etc. The most notable change is that `ELEKTRA_UNUSED`, `ELEKTRA_FALLTHROUGH` and `ELEKTRA_DEPRECATED` are now part of `kdbmacros.h` instead of `kdbconfig.h`. If you want I can extract that part into a separate PR, but nobody else seems to care whether `kdbconfig.h` is installed or not.
- We now install a CMake function called `elektra_kdb_gen` for invoking `kdb gen`. To make it easier to use, I added a new functionality to `kdb gen`. If it is called with the parent key `-`, it doesn't render any templates. Instead it prints the names of the files that would be written. This is needed so that CMake can automatically figure out which files depend on the `kdb gen` call.

markus2330 commented:
Amazing changes! Can you rebase and make the build server happy?

Jakob Fischer: Do you have time to look at it?

kodebach commented:
The `highlevel` template now also generates a shell script that can be used for mounting the specification. It most likely won't work out of the box, but it can be used as a starting point. In particular the path to the installed application (or spec file, if an external spec is used), might be wrong.

markus2330 commented:
Via autoconf we somehow need to deliver the correct SPEC_FILE. Can you please create an LCDproc issue so that we do not forget about this?

kodebach commented:
> Can you please create an LCDproc issue so that we do not forget about this?

Don't worry I won't forget. In fact I am already working on it. I am also looking into creating an autoconf version of the new `elektra_kdb_gen` CMake function.

bauhaus93 commented on doc/help/elektra-highlevel-gen.md:
```diff
@@ -0,0 +1,410 @@
+The main difference between allocating and non-allocating structs, is how their getter function works.
+
+Allocating structs use a getter similar to the one primitive types, strings and enums use. It returns a pointer to a newly allocated struct,
+which has to be free using the generated `ELEKTRA_STRUCT_FREE` function.
```
```suggestion
which has to be freed using the generated `ELEKTRA_STRUCT_FREE` function.
```

bauhaus93 commented on doc/tutorials/code-generator.md:
```diff
@@ -0,0 +1,272 @@
+[here](https://mustache.github.io/mustache.5.html). All of features supported by the kainjow library should be supported by our framework
+as well.
+
+Our CMke script will collect all `.mustache` files in `src/tools/kdb/gen/templates` into a header containing a `static const char *` field
```
```suggestion
Our CMake script will collect all `.mustache` files in `src/tools/kdb/gen/templates` into a header containing a `static const char *` field
```

markus2330 replied:
Thank you for the suggestions!

Did you also try everything in the tutorials? Did everything work?

bauhaus93 replied:
Not yet, but I will check them later this day.

kodebach commented:
jenkins build libelektra please

kodebach commented:
jenkins build libelektra please

kodebach commented:
@markus2330 Why can't Jenkins build this PR anymore? It complains about a non-empty directory. Any idea what could have caused that?

markus2330 commented:
Simply create a new PR. This also fixed a similar problem I had with the buster builds (#2876 did not work).

Did you change something in the Jenkinsfile? I suspected that this was the reason for #2876.

kodebach commented:
reopend as #2927
